common4j/src/main/com/microsoft/identity/common/java/base64/Base64.kt

```diff
@@ -0,0 +1,202 @@
+import java.lang.IllegalStateException
+import java.nio.charset.StandardCharsets
+
+/**
+ * A class consolidating all Base64 operations in our code base.
+ **/
+class Base64Util {
+    companion object {
+        private val TAG = Base64Util::class.java.simpleName
+
+        private var base64: IBase64 = initialize()
+
+        /**
+         * In Android, we'll use Android's own Base64 implementation.
+         * In other places, we'll use Msebera's Base64.
+         *
+         * Both are forked from a legacy version of apache commons-codec.
+         * (we had to use Msebera instead because the ClassLoader would throw an error due to a conflict between
+         * the classes in commons-codec library and the one in the OS.)
+         *
+         * Msebera library is no longer maintained, and it was recently flagged with an Android apk tool.
+         * Even if the flagged part is NOT the code path we use.
+         * It's better that we migrate away from it (so that our customer don't see the false alarm).
+         *
+         * .. so we put them in LinuxBroker and test projects instead.
+         **/
+        fun initialize() : IBase64 {
+            val androidBase64 = tryLoadAndroidBase64()
+            if (androidBase64 != null){
+                return androidBase64
+            }
+
+            val linuxBase64 = tryLoadMseberaBase64InLinux()
+            if (linuxBase64 != null){
+                return linuxBase64
+            }
+
+            val common4jUnitTestBase64 = tryLoadMseberaBase64InCommon4jUnitTest()
+            if (common4jUnitTestBase64 != null){
+                return common4jUnitTestBase64
+            }
+
+            val broker4jUnitTestBase64 = tryLoadMseberaBase64InBroker4jUnitTest()
+            if (broker4jUnitTestBase64 != null){
+                return broker4jUnitTestBase64
+            }
+
+            val testUtilsBase64 = tryLoadMseberaBase64InTestUtils()
+            if (testUtilsBase64 != null){
+                return testUtilsBase64
+            }
+
+            throw IllegalStateException("Cannot find a Base64 to initialize.")
+        }
+
+        private fun tryLoadAndroidBase64(): IBase64? {
+            return try {
+                val androidBase64 = Class.forName("com.microsoft.identity.common.base64.AndroidBase64").getDeclaredConstructor().newInstance() as IBase64
+
+                // If executed in Android Unit tests, androidBase64 will fail (mocking required) with a RuntimeException.
+                androidBase64.encode(ByteArray(0), Base64Flags.DEFAULT)
+
+                return androidBase64
+            } catch (e: ClassNotFoundException) {
+                null
+            } catch (e: RuntimeException){
+                null
+            }
+        }
+
+        private fun tryLoadMseberaBase64InLinux(): IBase64? {
+            return try {
+                Class.forName("com.microsoft.identity.broker.base64.MseberaBase64ForLinux")
```

Comment:
I see the intent of this class (to be able to use an all-encompassing Base64 Util class across the libraries, without the developer needing to know about the differences in dependencies underneath). But the use of hardcoded class name strings here seems interesting to me, especially when a broker class is being referenced in a common4j class.

I'm wondering if it would be possible to not have this consolidated Util class but instead just have each library use the designated class directly:
- Make the common4j, broker4j, testutils Msebera classes have the 'default' access modifier so that they are only accessible from within common4j/broker4j, and common/msal/broker cannot access them
- Keep the AndroidBase64 one the way it is (someone contributing to the Linux broker shouldn't be using a class which has Android in the name)
- Keep the Linux class the way it is

I'm additionally wondering if we could still name each of these classes Base64Util if they reside in different packages.
Alternatively, maybe some sort of Factory + depedency injection solution could be used?

Reply:
- I removed msebera from the entire android dependency chain because it got flagged.
(So msebera is no longer shipped with common4j)

Reply:
Lmk if that answers the question. We can have a call if preferred.

Reply:
Had an offline discussion; agreed to put these class strings in constants.